backend/lucene/src/main/java/org/hibernate/search/backend/lucene/index/impl/ShardingStrategyInitializationContextImpl.java

```diff
@@ -129,22 +127,18 @@ private void contributeShardWithSilentFailure(Map<String, Shard> shardCollector,
 								.withFallback( indexPropertySource )
 						: indexPropertySource;
 
-		DirectoryHolder directoryHolder = null;
 		try ( BeanHolder<? extends DirectoryProvider> directoryProviderHolder =
 				DIRECTORY_TYPE.getAndTransform( shardPropertySource, startContext.beanResolver()::resolve ) ) {
 			DirectoryCreationContext context = new DirectoryCreationContextImpl( shardEventContext,
 					indexName(), shardId, beanResolver(), shardPropertySource.withMask( "directory" ) );
-			directoryHolder = directoryProviderHolder.get().createDirectoryHolder( context );
 
 			IOStrategy ioStrategy = backendContext.createIOStrategy( shardPropertySource );
 
-			Shard shard = backendContext.createShard( model, shardEventContext, directoryHolder, ioStrategy,
-					shardPropertySource );
+			Shard shard = backendContext.createShard( model, shardEventContext, directoryProviderHolder.get(), context,
```

Comment:
Passing and saving the context in the shard/accessor seems very dodgy... most `context` objects are used once and then thrown away, and I'm a bit afraid of what unintended consequences storing a context could have. It could lead to unnecessary memory usage, for example, if the context internally holds a reference to something that is only necessary on startup.

Can't you just delay everything related to the `ioStrategy` (creating the index accessor, ...) until the `preStart` methods are called, just after we create the directory holder? Then you don't need to save the context anymore, you don't need to introduce `start` methods in `IndexReaderProvider`/`IndexWriterProvider` anymore, ...

Delaying all this would make sense regardless of your changes: in a hypothetical Quarkus + Hibernate Search + Lucene extension (or a Quarkus + Infinispan Embedded + Lucene extension), this would allow tuning index writers/readers at runtime, which is definitely something useful.

Also, I thought that was what we mentioned on Zulip initially... ?

If you're uncomfortable with making this change yourself, I can have a look and send a PR, so that you can rebase your changes on that.